src/Signer/Factory.php

```diff
@@ -0,0 +1,152 @@
+<?php
+/**
+ * This file is part of Lcobucci\JWT, a simple library to handle JWT and JWS
+ *
+ * @license http://opensource.org/licenses/BSD-3-Clause BSD-3-Clause
+ */
+
+namespace Lcobucci\JWT\Signer;
+
+use Lcobucci\JWT\Signer;
+
+/**
+ * Class that holds a registry of signers
+ *
+ * @author Woody Gilk <@shadowhand>
+ * @since 3.0.6
+ */
+class Factory
+{
+    /**
+     * The currently registered signers
+     *
+     * @var array
+     */
+    private $registry = [];
+
+    /**
+     * Prefixes used for namespaces
+     *
+     * @var array
+     */
+    private $prefixes = [
+        'ES' => 'Ecdsa',
+        'HS' => 'Hmac',
+        'RS' => 'Rsa',
+    ];
+
+    /**
+     * Initializes the factory, registering custom signers
+     *
+     * @param array $registry
+     * @param array $prefixes
+     */
+    public function __construct(array $registry = [], array $prefixes = [])
+    {
+        foreach ($registry as $signer) {
+            $this->register($signer);
+        }
+
+        $this->prefixes = array_replace($this->prefixes, $prefixes);
+    }
+
+    /**
+     * Register a signer
+     *
+     * @param Signer $signer
+     *
+     * @return self
+     */
+    public function register(Signer $signer)
+    {
+        $this->registry[$signer->getAlgorithmId()] = $signer;
+
+        return $this;
+    }
+
+    /**
+     * Get the namespace root for an algoritm id
+     *
+     * @param string $algoritm
+     *
+     * @return string
+     */
+    protected function getNamespace()
+    {
+        return __NAMESPACE__;
+    }
+
+    /**
+     * Get the namespace prefix for an algoritm id
+     *
+     * @param string $algoritm
+     *
+     * @return string|null
+     */
+    protected function getPrefix($algoritm)
+    {
+        $prefix = substr($algoritm, 0, 2);
+
+        if (empty($this->prefixes[$prefix])) {
```

Comment:
This is slightly weird because the method is `protected` but `$prefixes` is private. The intent is that this method would be overloaded to do a custom prefix check and then use `parent::getPrefix($algorithm)` if necessary.